### ISSUE TITLE
Downloading files on headless chrome

### ISSUE THREAD

**Issue by emosGambler** [angular/protractor#5245]
Hello people,

Honestly no solution from the web was working for me. In our company, we run protractor with headless chrome. The framework we use is serenity js.

Many tests depend on validating excel files, which should be downloaded by clicking button on UI. Unfortunately we receive all sorts of time outs while trying so. The file is not downloaded and just for the record, when I turn off the headless mode, it all works as expected.

Unfortunately I cannot give you any code examples, as the code belongs to my client. Although I tried these already:

1) Clicking the button directly
2) Going to download URL directly
3) Trying to download it by third party lib like request

I was thinking about running a new non-headless instance chrome just for the sake of downloading this file, but I had no luck in finding a proper way to do it.

Anyone had this problem and solved it? I know that without code example that could be difficult to solve, but I am out of ideas...

Thanks and regards.

**Comment by RainbowHorn** [comment]
As far as i know, this is not protractor-specific bug, but Chrome bug. You can see it this. On the same page you can find workarounds for different languages and framework, but basically for protractor solution is to send POST request directly to driver. Something like that, i think
```
browser.driver.sendChromiumCommand('Page.setDownloadBehavior', {
    'behavior': 'allow',
    'downloadPath': DOWNLOAD_DIR
});
```
(or like in this [comment](https://bugs.chromium.org/p/chromium/issues/detail?id=696481#c196))

Some comments says, that it's no longer works with some of the newer versions of Chrome but i think it's worth to try. 

Also, you can try to use containerized Chrome, which works headlessly without explicitly specifying ‘headless’ option. It's works for some people.

**Comment by cnishina** [comment]
Does this work on Firefox? Or is this Chrome specific? If it doesn't do it for both, could you instead get the link produced by the button and download the file with node?

**Comment by emosGambler** [comment]
> Does this work on Firefox? Or is this Chrome specific? If it doesn't do it for both, could you instead get the link produced by the button and download the file with node?

Hey @cnishina ! Tried that, no success

**Comment by cnishina** [comment]
So this sounds like a limitation of WebDriver rather than the Protractor framework.

**Comment by urvashidodeja** [comment]
no success with this

**Comment by abhinaba1080** [comment]
@urvashidodeja 

If you just only clicking the button to download a file, that should not cause timeout until and unless you are are verifying the download is successful or not. 

Now, sometimes chrome throws some pop up to help the user while downloading the file. You can set the right capabilities to tell chrome to download the file automatically into a directory. 

Try with below capabilities:

```ts
chrome: {
    browserName: 'chrome',
    platform: 'ANY',
    version: 'ANY',
    'goog:chromeOptions': {
      w3c: false,
      useAutomationExtension: false,
      excludeSwitches: ['enable-automation'],
      args: [
         '--headless',
        '--disable-gpu',
        '--test-type=browser',
        '--disable-extensions',
        '--no-sandbox',
        '--disable-infobars',
        '--window-size=1920,1080'
      ],
      prefs: {
        credentials_enable_service: false,
        profile: {
          password_manager_enabled: false
        },
        download: {
          prompt_for_download: false,
          directory_upgrade: true,
          default_directory: process.cwd() + `\\downloads\\chrome`
        }
      }
    }
```

**Comment by urvashidodeja** [comment]
> @urvashidodeja
> 
> If you just only clicking the button to download a file, that should not cause timeout until and unless you are are verifying the download is successful or not.
> 
> Now, sometimes chrome throws some pop up to help the user while downloading the file. You can set the right capabilities to tell chrome to download the file automatically into a directory.
> 
> Try with below capabilities:
> 
> ```ts
> chrome: {
>     browserName: 'chrome',
>     platform: 'ANY',
>     version: 'ANY',
>     'goog:chromeOptions': {
>       w3c: false,
>       useAutomationExtension: false,
>       excludeSwitches: ['enable-automation'],
>       args: [
>          '--headless',
>         '--disable-gpu',
>         '--test-type=browser',
>         '--disable-extensions',
>         '--no-sandbox',
>         '--disable-infobars',
>         '--window-size=1920,1080'
>       ],
>       prefs: {
>         credentials_enable_service: false,
>         profile: {
>           password_manager_enabled: false
>         },
>         download: {
>           prompt_for_download: false,
>           directory_upgrade: true,
>           default_directory: process.cwd() + `\\downloads\\chrome`
>         }
>       }
>     }
> ```

@abhinaba1080 I eventually have to verify the success of download . The file is not getting downloaded in headless mode. And hence i am not able to get count, whereas in non-headless mode, file gets downloaded and count is also verified. I am struggling with verifying download in headless mode using protractor

**Comment by abhinaba1080** [comment]
Yes. so first try to check manually whether the above capability can down the file in root > downloads > chrome folder or not. For verification of the download you can try below:

```ts
const verifyFileDownloaded = async (fileName: string) => {
  await logger.info('Getting users download path: ' + (await this.downloadsFolder()));
  const filePath: string = ( process.cwd() + "\\downloads\\chrome" + '\\' + fileName).replace(/\\/g, '/');
  await logger.info('Getting the path ' + filePath);
  return browser.wait(async () => fs.existsSync(filePath), Number(process.env.IMPLICIT_WAIT), 'File never appeared!');
};

```

Just call this method after clicking the download button to verify the file is downloaded.

**Comment by urvashidodeja** [comment]
@abhinaba1080 I tried the above chrome configs, i am still not able to download file. I am not getting file downloaded in my directory.

**Comment by abhinaba1080** [comment]
Can you Paste the error message you are getting?

**Comment by urvashidodeja** [comment]
@abhinaba1080  I am not getting any error. 
4 specs, 0 failures
Finished in 46.806 seconds

0
Its just that File isnt getting downloaded even checked the folder and count is 0 .
When i do the same without headless, file gets downloaded in folder and count i get is 1

**Comment by abhinaba1080** [comment]
Can you share below:

1. Protractor conf file
2. The steps where you are clicking the button
3. The verification method

**Comment by awhisler** [comment]
Does anyone have code actually opening the downloaded file?
